backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java

```diff
@@ -38,38 +39,49 @@
   private ProviderRepository _providerRepo;
   private AuthorizationService _authService;
   private OktaRepository _oktaRepo;
+  private CurrentOrganizationRolesContextHolder _currentOrgRolesContextHolder;
 
   public OrganizationService(
       OrganizationRepository repo,
       FacilityRepository facilityRepo,
       AuthorizationService authService,
       ProviderRepository providerRepo,
-      OktaRepository oktaRepo) {
+      OktaRepository oktaRepo,
+      CurrentOrganizationRolesContextHolder currentOrgRolesContextHolder) {
     _repo = repo;
     _facilityRepo = facilityRepo;
     _authService = authService;
     _providerRepo = providerRepo;
     _oktaRepo = oktaRepo;
+    _currentOrgRolesContextHolder = currentOrgRolesContextHolder;
   }
 
   public Optional<OrganizationRoles> getCurrentOrganizationRoles() {
-    List<OrganizationRoleClaims> orgRoles = _authService.findAllOrganizationRoles();
-    List<String> candidateExternalIds =
-        orgRoles.stream()
-            .map(OrganizationRoleClaims::getOrganizationExternalId)
-            .collect(Collectors.toList());
-    List<Organization> validOrgs = _repo.findAllByExternalId(candidateExternalIds);
-    if (validOrgs == null || validOrgs.size() != 1) {
-      int numOrgs = (validOrgs == null) ? 0 : validOrgs.size();
-      LOG.warn("Found {} organizations for user", numOrgs);
-      return Optional.empty();
-    }
-    Organization foundOrg = validOrgs.get(0);
-    Optional<OrganizationRoleClaims> foundRoles =
-        orgRoles.stream()
-            .filter(r -> r.getOrganizationExternalId().equals(foundOrg.getExternalId()))
-            .findFirst();
-    return foundRoles.map(r -> getOrganizationRoles(foundOrg, r));
+    return _currentOrgRolesContextHolder
+        .getOrganizationRoles()
+        .or(
```

Comment:
I have a couple thoughts about this, one micro and one macro.

Micro first: it seems weird to me to have the `Supplier` surreptitiously make a `set` call to the thing we are getting the underlying Optional from. It seems more logical (to me, anyway) to separate out the actual fetching code from the context-manipulation code:

```java
if (found.isEmpty()) {
    Optional<OrganizationRoles> r = getActualOrganizationRoles()
    _currentOrgRolesContextHolder.setOrganizationRoles(r);
   found = r;
}
return found;
```

More macro question: is it allowed at this point to have a `null` OrganizationRoles in any context where we might call this? I'm pretty thoroughly swapped out of this part of my brain right now so I'm not sure, but it seems like there are either two states ("not yet called in this request"; "previously called and we stored a valid OrganizationRoles") or three states ("not yet called"; "previously called and we stored a valid OrganizationRoles"; "previously called but the user has no Organization role").

My gut feeling is that one or both of patient experience and admin login is capable of producing that condition, in which case we need to be able to detect the difference between the last two states.

Reply:
Yeah, that's a fair concern about the `Supplier`. My original code was actually _more_ along those lines but I refactored it to use the `Optional.or` to satisfy a SonarQube complaint about a `null`-valued Optional. But you're right, that set inside the Supplier smells off.

Re: your macro consideration: I don't know, because this method already returned an Optional before I got in here and I didn't want to mess with it. I imagine that was done naively just extending out from the fact that the Repository's `find` method returns an Optional, but I would rather table that whole conversation for another day because it's outside of the more focused scope of this changeset.

Reply:
Done: fa02f80c

Reply:
Macro concern done in c3d186be